File: Hard/Poker_Hand.py
''' You are playing poker with your friends and need to evaluate your hand.  
A hand consists of five cards and is ranked, from lowest to highest, in the following way: 
High Card: Highest value card (from 2 to Ace). 
One Pair: Two cards of the same value. 
Two Pairs: Two different pairs. 
Three of a Kind: Three cards of the same value. 
Straight: All cards are consecutive values. 
Flush: All cards of the same suit. 
Full House: Three of a kind and a pair. 
Four of a Kind: Four cards of the same value. 
Straight Flush: All cards are consecutive values of same suit. 
Royal Flush: 10, Jack, Queen, King, Ace, in same suit.  
 
Task: 
Output the rank of the give poker hand.  
 
Input Format:  
A string, representing five cards, each indicating the value and suite of the card, separated by spaces.  
Possible card values are: 2 3 4 5 6 7 8 9 10 J Q K A 
Suites:  H (Hearts), D (Diamonds), C (Clubs), S (Spades) 
For example, JD indicates Jack of Diamonds.  
 
Output Format: 
A string, indicating the rank of the hand (in the format of the above description).  
 
Sample Input: 
JS 2H JC AC 2D 
 
Sample Output:  
Two Pairs

Explanation: The hand includes two Jacks and two 2s, resulting in Two Pairs. '''

import sys
myHand = input()

myDict = {
    '2': 1,
    '3': 2,
    '4': 3,
    '5': 4,
    '6': 5,
    '7': 6,
    '8': 7,
    '9': 8,
    '10': 9,
    'J': 10,
    'Q': 11,
    'K': 12,
    'A': 13
}

myHandList = myHand.split()

# checking Flush (if myHandSuite = 5)
if len(myHandList[0]) == 2:
    myHandSuite = myHand.count(myHand[1])
else:
    myHandSuite = myHand.count(myHand[2])

# creating the list of values    
myValueList = []
for i in myHandList:
    if len(i) == 2:
        myValueList.append(i[0:1])
    else:
        myValueList.append(i[0:2])

# sorting myValueList
myValueListSorted = sorted(myValueList, key = lambda item: myDict[item])

# checking Straight (if straightCheck = 5)
straightCheck = 0
i = 0
for j in myValueListSorted:
    if myDict.get(myValueListSorted[i]) == myDict.get(myValueListSorted[i + 1]) - 1:
        straightCheck += 1
        if i < 3:
            i += 1

if myHandSuite == 5:
    if straightCheck == 5 and "A" in myValueList:
        print("Royal Flush")
        sys.exit()
    if straightCheck == 5 and "A" not in myValueList:
        print("Straight Flush")
        sys.exit()
    else: 
        print("Flush")
        sys.exit()
        
if myHandSuite != 5 and straightCheck == 5:
    print("Straight")
    sys.exit()
'''   
num = [myValueListSorted[0].count]
j = 1
while j < 5:
    if myValueListSorted[j] != myValueListSorted[j - 1]
        num.append(myValueListSorted[j].count)
'''    
    
if  myValueListSorted.count(myValueListSorted[0]) == 4 or myValueListSorted.count(myValueListSorted[1]) == 4:
    print("Four of a Kind")
    sys.exit()
    
if  (myValueListSorted.count(myValueListSorted[0]) == 2 and myValueListSorted.count(myValueListSorted[2]) == 3) or (myValueListSorted.count(myValueListSorted[0]) == 3 and myValueListSorted.count(myValueListSorted[3]) == 2):
    print("Full House")
    sys.exit()

if  myValueListSorted.count(myValueListSorted[0]) == 3 or myValueListSorted.count(myValueListSorted[1]) == 3 or myValueListSorted.count(myValueListSorted[2]) == 3:
    print("Three of a Kind")
    sys.exit()        

if  myValueListSorted.count(myValueListSorted[0]) == 2 and myValueListSorted.count(myValueListSorted[2]) == 2:
    print("Two pairs")
    sys.exit()  
if  myValueListSorted.count(myValueListSorted[1]) == 2 and myValueListSorted.count(myValueListSorted[3]) == 2:
    print("Two pairs")
    sys.exit() 
if  myValueListSorted.count(myValueListSorted[0]) == 2 and myValueListSorted.count(myValueListSorted[3]) == 2:
    print("Two pairs")
    sys.exit() 

k = 0
for i in myValueListSorted:
    OnePair = myValueListSorted.count(myValueListSorted[k])
    k += 1
    if OnePair == 2:
        print("One Pair")
        sys.exit() 
        
print("High Card")

File: Medium/Hard/Poker_Hand.c
/* You are playing poker with your friends and need to evaluate your hand.  
A hand consists of five cards and is ranked, from lowest to highest, in the following way: 
High Card: Highest value card (from 2 to Ace). 
One Pair: Two cards of the same value. 
Two Pairs: Two different pairs. 
Three of a Kind: Three cards of the same value. 
Straight: All cards are consecutive values. 
Flush: All cards of the same suit. 
Full House: Three of a kind and a pair. 
Four of a Kind: Four cards of the same value. 
Straight Flush: All cards are consecutive values of same suit. 
Royal Flush: 10, Jack, Queen, King, Ace, in same suit.  
 
Task: 
Output the rank of the give poker hand.  
 
Input Format:  
A string, representing five cards, each indicating the value and suite of the card, separated by spaces.  
Possible card values are: 2 3 4 5 6 7 8 9 10 J Q K A 
Suites:  H (Hearts), D (Diamonds), C (Clubs), S (Spades) 
For example, JD indicates Jack of Diamonds.  
 
Output Format: 
A string, indicating the rank of the hand (in the format of the above description).  
 
Sample Input: 
JS 2H JC AC 2D 
 
Sample Output:  
Two Pairs

Explanation: The hand includes two Jacks and two 2s, resulting in Two Pairs. */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
int main(){
    
    char input[100];  
    fgets(input,100,stdin);
    int length = strlen(input);
    char *values[]={"2","3","4","5",           
                    "6","7","8","9",
              "10","J","Q","K","A"};
    int count[13]={0};
//////////////////////// finding suit    
    char rev_input[100]={0};
    for(int i=0; i<length; ++i){
        rev_input[i]= input[length-1-i];
    }
    int count_suit=1;                    
    for(int i=1; i<length; ++i){
       if(rev_input[0]==rev_input[i])   
          ++count_suit; 
         //should be 5 for Flush
    }
///////////////////// counting cards   
    for(int i=0; i<length; ++i){
        for(int j=0; j<13;++j){
            if(input[i]==*values[j]){
                ++count[j];
            } 
        }
    }
//////////////////// finding straight 
    int count_straight = 1;
    for(int i=0; i<13;++i){
        if (count[i]==1){
            while(count[++i] == 1){
                 ++count_straight; 
       //should be 5 for Straight
            }
            break;
        }        
    }   
/////////////////////////////////////
    if (count_straight==5 && count[12]==1 && count_suit==5){
        printf("Royal Flush");
        return 0;        
    }
    if (count_straight==5 && count_suit==5){
        printf("Straight Flush");
        return 0;        
    }
    if (count_straight==5){
        printf("Straight");
        return 0;        
    }
    if (count_suit==5){
        printf("Flush");
        return 0;        
    }
/////////////////////////  finding 4    
    for(int i=0; i<13; ++i){
        if (count[i]==4){
            printf("Four of a Kind");
            return 0;
        }
/////////////////  finding 3 + 2 or 3
        if (count[i]==3){
           int m=i;
           while (++m<=12){
             if (count[m]==2){
                printf("Full House");
                return 0;
                }
           }
           printf("Three of a Kind");
           return 0;
        }
////////  finding 2 + 3 or 2 + 2 or 2
        if (count[i]==2){
            int n=i;
            while (++n<=12){
              if (count[n]==3){
                printf("Full House");
                return 0;
              }
            }
            while (++n<=12){
              if (count[n]==2){
                printf("Two Pairs");
                return 0;
              }
            }
            printf("One Pair");
            return 0;   
        }
    }
    printf("High Card");
    return 0;
}
